### ISSUE TITLE
FacebookBot startTicking is not called unless setupWebserver is called

### ISSUE THREAD

**Issue by smalltalk-josh** [howdyai/botkit#772]
I am using botkit with Facebook Messenger, but I am not calling `setupWebserver`. My setup is similar to [using an Express web server alongside BotKit for Facebook Messenger](https://github.com/howdyai/botkit/blob/master/docs/readme-facebook.md#use-botkit-for-facebook-messenger-with-an-express-web-server), but I do call `createWebhookEndpoints`.

The biggest issue I noticed was that conversation did not work at all. See: [#503 - convo.say not working](https://github.com/howdyai/botkit/issues/503#issuecomment-270219099)

After digging around, I see that `startTicking` was not called, because I never called `setupWebserver`. This is a bug, since `startTicking` is a core to the bot, but `setupWebserver` is optional.

I am going to submit a PR, but the different bots handle calling `startTicking` differently.

`startTicking` - when called:

- ConsoleBot - [inside `text_botkit.listenStdIn`](https://github.com/howdyai/botkit/blob/master/lib/ConsoleBot.js#L72)
- BotFramework - [inside `bf_botkit.createWebhookEndpoints`](https://github.com/howdyai/botkit/blob/b13e5de9e208c7a76d59bf194ffd76f82a4e398c/lib/BotFramework.js#L156)
- Twiliobot - [inside `twilio_botkit.createWebhookEndpoints`](https://github.com/howdyai/botkit/blob/e00448dd2ee7d1bd048ebbfb3f3f3dd8f4e59026/lib/TwilioIPMBot.js#L221)
- Tropo - [inside ctor `function Tropobot(configuration)`](https://github.com/howdyai/botkit/blob/5984014fb5a988d2dd6e0bd47259d9f2c0a5e11b/lib/Tropo.js#L214)
- Sparkbot - [inside ctor `function Sparkbot(configuration))`](https://github.com/howdyai/botkit/blob/0188c5bfce2c7bf1942234d9aa057cc1c588ee61/lib/CiscoSparkbot.js#L477)
- Slackbot_worker - [inside - `bot.rtm.on('open')`](https://github.com/howdyai/botkit/blob/34b44646e95d0abff7dbc96d8a9acbce36f11878/lib/Slackbot_worker.js#L215)
- Facebook - [inside - `facebook_botkit.setupWebserver`](https://github.com/howdyai/botkit/blob/3b175b68cc5cdf6ae7e9d075cfefcd697a8ed965/lib/Facebook.js#L417)

My takeaways:
- Facebook - it is a bug #503 and my experiences; it should be in createWebhookEndpoints or in ctor
- Slackbot_worker - might be a bug: #737 

I am going to take a shot at the Facebook issue. Anybody know of a reason to move or not move the call to `startTicking` in Slackbot_worker?

**Comment by jonchurch** [comment]
`createWebhookEndpoints()` is also optional, this is not a bug in my eyes.

You can call `controller.startTicking()` from your main bot.js file (or anywhere else), as is the convention with Facebook bots in Botkit. 

When using your own express server with botkit, the emphasis is on flexibility and decoupling. It would be a code smell to me personally to pass my webserver into `controller.createWebhookEndpoints()` and that is not how it is done in the example starter kits which decouple express from the controller as much as possible. Botkit has created those convenience methods for folks who don't want to setup their own server. 

Because of how important `startTicking` is to your bot, I recommend putting it in your main bot file, the logic reads better in my opinion.

tldr: Don't use createWebhookEndpoints if you aren't using setupWebserver, call startTicking from your main bot.js file.

**Comment by smalltalk-josh** [comment]
OK, it just seems odd that FacebookBot the only bot type that doesn't call `startTicking` inside the `ctor` or `createWebhookEndpoints()`

**Comment by jonchurch** [comment]
You are correct that is weird! I'm just trying to help the maintainers of Botkit cut down on extra work by being a harsh defender of code changes. Also this _could_ be breaking for some.

These are my opinions, I'm just a contributor like you, there is a fine chance the maintainers may realize this discrepancy and want to fix it.

My goal is to make sure people who run into these issues from google get the correct info, as custom express server setup is one of the most commonly asked questions here and in the dev4slack developer channel.

There's a sordid history with startTicking. I've scratched my head before thinking about where it should go. It used to get called for you when connecting a bot to Slack via RTM, then Events API came along, and suddenly it wasn't being called at all, so we stuck it into createWebhookEndpoints to help people out who were creating express servers that way, then quite quickly after that convention shifted to this fancy schmancy [more extensible way to structure botkit projects](https://github.com/howdyai/botkit-starter-slack) and it was being called as part of the main bot.js file.

**Comment by smalltalk-josh** [comment]
@jonchurch 👍 great history/background. thx for insight

**Comment by jonchurch** [comment]
@smalltalk-josh No worries bud! I sincerely hope you submit more fixes to botkit! It took a good eye to catch that (well and stepping in it like that haha) and the due diligence on all the other bots has brought up a very valid point, startTicking is all over the place.

Now the community can figure out what/if to do, because of your work!

**Comment by jonchurch** [comment]
I would update the PR with your findings above about the other bots, btw. Useful info that might get lost here
